engines/sherlock/resources.h

```diff
@@ -0,0 +1,128 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#ifndef SHERLOCK_RESOURCES_H
+#define SHERLOCK_RESOURCES_H
+
+#include "common/array.h"
+#include "common/file.h"
+#include "common/hashmap.h"
+#include "common/hash-str.h"
+#include "common/rect.h"
+#include "common/str.h"
+#include "common/stream.h"
+#include "graphics/surface.h"
+
+namespace Sherlock {
+
+typedef Common::Array<byte> CacheEntry;
+typedef Common::HashMap<Common::String, CacheEntry, Common::IgnoreCase_Hash, Common::IgnoreCase_EqualTo> CacheHash;
+
+struct LibraryEntry {
+	uint32 _offset, _size;
+	int _index;
+
+	LibraryEntry() : _index(0), _offset(0), _size(0) {}
+	LibraryEntry(int index, uint32 offset, uint32 size) :
+		_index(index), _offset(offset), _size(size) {}
+};
+typedef Common::HashMap<Common::String, LibraryEntry, Common::IgnoreCase_Hash, Common::IgnoreCase_EqualTo> LibraryIndex;
+typedef Common::HashMap<Common::String, LibraryIndex, Common::IgnoreCase_Hash, Common::IgnoreCase_EqualTo> LibraryIndexes;
+
+class SherlockEngine;
+
+class Cache {
+private:
+	SherlockEngine *_vm;
+	CacheHash _resources;
+public:
+	Cache(SherlockEngine *_vm);
+
+	bool isCached(const Common::String &filename) const;
+
+	void load(const Common::String &name);
+	void load(const Common::String &name, Common::SeekableReadStream &stream);
+
+	Common::SeekableReadStream *get(const Common::String &filename) const;
+};
+
+class Resources {
+private:
+	SherlockEngine *_vm;
+	Cache _cache;
+	LibraryIndexes _indexes;
+	int _resourceIndex;
+
+	void loadLibraryIndex(const Common::String &libFilename, Common::SeekableReadStream *stream);
+public:
+	Resources(SherlockEngine *vm);
+
+	void addToCache(const Common::String &filename);
+	void addToCache(const Common::String &filename, const Common::String &libFilename);
+	void addToCache(const Common::String &filename, Common::SeekableReadStream &stream);
+	bool isInCache(const Common::String &filename) const { return _cache.isCached(filename); }
+
+	void decompressIfNecessary(Common::SeekableReadStream *&stream);
+
+	Common::SeekableReadStream *load(const Common::String &filename);
+
+	Common::SeekableReadStream *load(const Common::String &filename, const Common::String &libraryFile);
+
+	bool exists(const Common::String &filename) const;
+
+	int resourceIndex() const;
+
+	static Common::SeekableReadStream *decompressLZ(Common::SeekableReadStream &source, uint32 outSize);
+	Common::SeekableReadStream *decompressLZ(Common::SeekableReadStream &source);
+};
+
+struct ImageFrame {
+	uint32 _size;
+	uint16 _width, _height;
+	int _paletteBase;
+	bool _rleEncoded;
+	Common::Point _offset;
+	byte _rleMarker;
+	Graphics::Surface _frame;
+
+	operator Graphics::Surface &() { return _frame; }
```

Comment:
I am not really a huge fan of implicit conversion operators, seeing that this also seems to duplicate the width/height values both in ImageFrame and ImageFrame::_frame I wonder if subclassing Surface might be worth considering (of course this has its own fair share of new issues).

What is the real reason to use implict conversion here? To prevent typing `._frame` or anything more sophisticated?